### PR TITLE
Adding support for typed get for tuple, e.g. get<typename T>(tuple<Ts...>& t)

### PR DESCRIPTION
**Pull request by CypherSignal**
These commits add the typed "get" that was added for tuple in c++14, e.g. items (5)-(7) as described here: http://en.cppreference.com/w/cpp/utility/tuple/get

A couple tests of this are left in in-line with some other simple base tests for the tuple, as well as a small battery of tests specifically for checking some edge cases and each overload of the function.

The "tuple_index" helper was implemented similarly to "tuple_element". It calculates the index by iterating down to the type in question being 'found' in the template list, counting up the index value as it goes. The duplicatetypecheck is used to verify that we have not hit a type before; this allows for protection against a typed access to an ambiguous TupleLeaf. For example, with "tuple<double, double, bool> aTupleWithRepeatedType;", if "get<double>(aTupleWithRepeatedType)" is entered in, the follwing compilation errors will come up:

> 1>C:\Users\Cypher\Coding Projects\EASTL\include\EASTL/tuple.h(166): error C2338: duplicate type T in tuple_vector::get<T>(); unique types must be provided in declaration, or only use get<size_t>()
> 1>  C:\Users\Cypher\Coding Projects\EASTL\include\EASTL/tuple.h(180): note: see reference to class template instantiation 'eastl::tuple_index<T,eastl::TupleTypes<double,double,bool>>' being compiled
> 1>          with
> 1>          [
> 1>              T=double
> 1>          ]
> 1>  C:\Users\Cypher\Coding Projects\EASTL\include\EASTL/tuple.h(458): note: see reference to class template instantiation 'eastl::tuple_index<T,eastl::Internal::TupleImpl<eastl::integer_sequence<std::size_t,0,1,2>,double,double,bool>>' being compiled
> 1>          with
> 1>          [
> 1>              T=double
> 1>          ]
> 1>  C:\Users\Cypher\Coding Projects\EASTL\include\EASTL/tuple.h(805): note: see reference to function template instantiation 'T &eastl::Internal::get<T,eastl::integer_sequence<std::size_t,0,1,2>,double,double,bool>(eastl::Internal::TupleImpleastl::integer_sequence<std::size_t,0,1,2,double,double,bool> &)' being compiled
> 1>          with
> 1>          [
> 1>              T=double
> 1>          ]
> 1>  C:\Users\Cypher\Coding Projects\EASTL\test\source\TestTuple.cpp(215): note: see reference to function template instantiation 'T &eastl::get<double,double,double,bool>(eastl::tuple<double,double,bool> &)' being compiled
> 1>          with
> 1>          [
> 1>              T=double
> 1>          ]

Sadly, I don't know if there's a way to add a unit test of a guaranteed failure, to make sure that that _never_ compiles, so that has been omitted.

### CODE REVIEW

File: include/EASTL/tuple.h [hunk 1]
```diff
@@ -145,6 +145,41 @@ class tuple_element<I, const volatile Internal::TupleImpl<Indices, Ts...>> : pub
...
+template <typename T, typename... Ts>
+struct tuple_index<T, TupleTypes<T, Ts...>>
+{
+	typedef int DuplicateTypeCheck;
+	static_assert(is_void<typename tuple_index<T, TupleTypes<Ts...>>::DuplicateTypeCheck>::value, "duplicate type T in tuple_vector::get<T>(); unique types must be provided in declaration, or only use get<size_t>()");
+
+	static const size_t index = 0;
+};
+
+template <typename T, typename Ts, typename... TsRest>
```
Comment (rparolin) [inline include/EASTL/tuple.h]: I'd rename Ts to TsHead to clearly show that you are taking the head of the type list and then recursing further down the type list.  This will help to differentiate it from T which is the requested type.

File: include/EASTL/tuple.h [hunk 2]
```diff
@@ -145,6 +145,41 @@ class tuple_element<I, const volatile Internal::TupleImpl<Indices, Ts...>> : pub
...
+	typedef void DuplicateTypeCheck;
+	tuple_index() = delete; // tuple_index should only be used for compile-time assistance, and never be instantiated
+	static const size_t index = 0;
+};
+
+template <typename T, typename... Ts>
+struct tuple_index<T, TupleTypes<T, Ts...>>
+{
+	typedef int DuplicateTypeCheck;
+	static_assert(is_void<typename tuple_index<T, TupleTypes<Ts...>>::DuplicateTypeCheck>::value, "duplicate type T in tuple_vector::get<T>(); unique types must be provided in declaration, or only use get<size_t>()");
```
Comment (rparolin) [inline include/EASTL/tuple.h]: I'd leave a comment here explaining how we are marking the recursive template instantiations to ensure there are not duplicates.

**Comment by CypherSignal** [comment]
Commit e9ed2ab has the adjustments requested.

I feel like a part of me wanted to get that comment in anyway - if it's something I felt needed explaining in the pr, maybe it's something that needs explanation in line with the code :)

File: include/EASTL/tuple.h [hunk 3]
```diff
@@ -145,6 +145,44 @@ class tuple_element<I, const volatile Internal::TupleImpl<Indices, Ts...>> : pub
...
+	tuple_index() = delete; // tuple_index should only be used for compile-time assistance, and never be instantiated
+	static const size_t index = 0;
+};
+
+template <typename T, typename... TsRest>
+struct tuple_index<T, TupleTypes<T, TsRest...>>
+{
+	typedef int DuplicateTypeCheck;
+	// after finding type T in the list of types, try to find type T in TsRest.
+	// If we stumble back into this version of tuple_index, i.e. type T appears twice in the list of types, then DuplicateTypeCheck will be of type int, and the statiC_assert will fail.
```
Comment (rparolin) [inline include/EASTL/tuple.h]: small typo

Comment (rparolin) [inline include/EASTL/tuple.h]: "statiC_assert"

Comment (CypherSignal) [inline include/EASTL/tuple.h]: No wonder I didn't eyeball that; the difference between capital and lowercase in my ide's font in this case is really hard to see.
image

**Comment by rparolin** [comment]
Thanks for the PR.
